File: Volta/Volta/Calculations/DepreciationCalculator.swift
import Foundation

enum DepreciationCalculator {

    /// AfA-Basis = Gebäudewert + (Nebenkosten × Gebäudeanteil) + aktivierungspflichtige Renovierung
    static func afaBasis(
        buildingValue: Double,
        closingCostsTotal: Double,
        purchasePrice: Double,
        renovationAfaEligible: Double
    ) -> Double {
        guard purchasePrice > 0 else { return 0.0 }
        let buildingShareRatio = buildingValue / purchasePrice
        return buildingValue + (closingCostsTotal * buildingShareRatio) + renovationAfaEligible
    }

    /// Jährliche AfA
    static func depreciationYearly(afaBasis: Double, rate: Double) -> Double {
        afaBasis * rate
    }

    /// Monatliche AfA
    static func depreciationMonthly(afaBasis: Double, rate: Double) -> Double {
        depreciationYearly(afaBasis: afaBasis, rate: rate) / 12.0
    }

    /// AfA im Erwerbsjahr: anteilig ab erstem vollen Monat nach wirtschaftlichem Übergang.
    static func depreciationProratedInAcquisitionYear(
        afaBasis: Double,
        rate: Double,
        economicTransferDate: Date
    ) -> Double {
        let monthsRemaining = economicTransferDate.remainingMonthsInYear
        return depreciationMonthly(afaBasis: afaBasis, rate: rate) * Double(monthsRemaining)
    }
}

File: web/lib/calculations/depreciationCalculator.ts
import { monthOf } from './dateHelpers';

/**
 * AfA-Basis = Gebäudewert + (Nebenkosten × Gebäudeanteil) + aktivierungspflichtige Renovierung.
 * buildingValue and landValue come from the government valuation spreadsheet (Sachwertverfahren);
 * the building-share ratio is derived from them, never the other way around.
 */
export function afaBasis(
  buildingValue: number,
  closingCostsTotal: number,
  purchasePrice: number,
  renovationAfaEligible: number
): number {
  if (purchasePrice <= 0) return 0;
  const buildingShareRatio = buildingValue / purchasePrice;
  return buildingValue + closingCostsTotal * buildingShareRatio + renovationAfaEligible;
}

export function depreciationYearly(basis: number, rate: number): number {
  return basis * rate;
}

export function depreciationMonthly(basis: number, rate: number): number {
  return depreciationYearly(basis, rate) / 12;
}

/** Fractional deviation of (buildingValue + landValue) from purchasePrice; 0 when purchasePrice or both values are 0. */
export function valuationDeviation(buildingValue: number, landValue: number, purchasePrice: number): number {
  if (purchasePrice <= 0 || (buildingValue <= 0 && landValue <= 0)) return 0;
  return Math.abs(buildingValue + landValue - purchasePrice) / purchasePrice;
}

/**
 * AfA in the acquisition year: prorated from the first full month after
 * economicTransferDate (the month of transfer itself counts in full, per §7 EStG).
 */
export function depreciationProratedInAcquisitionYear(
  basis: number,
  rate: number,
  economicTransferDate: Date
): number {
  const monthsRemaining = 13 - monthOf(economicTransferDate); // e.g. Feb (2) -> 11 months (Feb-Dec)
  return depreciationMonthly(basis, rate) * monthsRemaining;
}
